backend/cfg/cfg_available_regs.ml

```diff
@@ -165,158 +156,159 @@ module Transfer = struct
+               || Reg.is_of_type_addr (RD.reg reg)
             then not remains_available
             else (
               instr.live <- Reg.Set.add (RD.reg reg) instr.live;
               false))
           avail_before
-      | false -> RD.Set.empty
+      | false -> RD_quotient_set.empty
     in
-    let made_unavailable = RD.Set.union made_unavailable_1 made_unavailable_2 in
-    let avail_across = RD.Set.diff avail_before made_unavailable in
+    let made_unavailable =
+      RD_quotient_set.union made_unavailable_1 made_unavailable_2
+    in
+    let avail_across = RD_quotient_set.diff avail_before made_unavailable in
     let avail_after =
-      (* If a result register will never be named, we can forget about it for
-         the purposes of this analysis. *)
-      let res = Reg.inter_set_array !all_regs_that_might_be_named instr.res in
-      RD.Set.union (RD.Set.without_debug_info res) avail_across
+      let res = Reg.set_of_array instr.res in
+      RD_quotient_set.union
+        (RD_quotient_set.without_debug_info res)
+        avail_across
     in
-    Some (ok avail_across), ok avail_after
+    dprintf "...avail_before %a\n%!" RD_quotient_set.print avail_before;
+    dprintf "...avail_across %a\n%!" RD_quotient_set.print avail_across;
+    dprintf "...avail_after %a\n%!" RD_quotient_set.print avail_after;
+    ok avail_across, ok avail_after
 
   let basic ({ avail_before } : domain) (instr : Cfg.basic Cfg.instruction) () :
       domain =
-    assert (Option.is_some avail_before);
+    dprintf "Instruction: (id=%a) %a\n%!" InstructionId.print instr.id
+      Cfg.print_basic instr;
     instr.available_before <- avail_before;
-    let avail_before = Option.get avail_before in
     if !Dwarf_flags.ddebug_invariants
     then check_invariants instr ~print_instr:Cfg.print_basic ~avail_before;
     let avail_across, avail_after =
       match avail_before with
-      | Unreachable -> None, unreachable
+      | Unreachable -> unreachable, unreachable
       | Ok avail_before -> (
         match instr.desc with
-        | Op
-            (Name_for_debugger
-              { ident; which_parameter; provenance; is_assignment; regs }) ->
-          (* First forget about any existing debug info to do with [ident] if
-             the naming corresponds to an assignment operation. *)
-          let forgetting_ident : RD.Set.t =
-            if not is_assignment
-            then avail_before
-            else
-              RD.Set.map
-                (fun reg ->
-                  match RD.debug_info reg with
-                  | None -> reg
-                  | Some debug_info ->
-                    if V.same (RD.Debug_info.holds_value_of debug_info) ident
-                    then RD.clear_debug_info reg
-                    else reg)
-                avail_before
-          in
-          let avail_after = ref forgetting_ident in
+        | Op (Name_for_debugger { ident; which_parameter; provenance; regs }) ->
```

Comment:
There is something subtle here about the handling of `Name_for_debugger` in  `Cfg_availability` : the field `regs` is part of the operation description (not instruction argument or result), and it starts with pseudo-registers. We rely on the mutability of `Reg.loc` to update these regs to physical locations. Right? If so, there is nothing that forces these registers to be updated throughout the backend, or preserve sharing of `Reg.t`. For example, register allocation can introduce new temporaries in the split phase and substitute them into instructions, leaving old temporaries in `Name_for_debugger` only. This could be another source of missing debug info.